hack/compare_states.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	rows := make(map[string]bool)
+	for _, row := range grid.GetRows() {
+		rows[row.GetName()] = true
+	}
+
+	columns := make(map[string]bool)
+	for _, column := range grid.GetColumns() {
+		columns[proto.MarshalTextString(column)] = true
+	}
+
+	return rows, columns, nil
+}
+
+func compare(ctx context.Context, first, second gcs.Path, client gcs.Client) (bool, error) {
+	var diffed bool
+	logrus.Infof("*****************************")
+	logrus.Infof("Comparing %q and %q...", first.String(), second.String())
+	firstRows, firstColumns, err := rowsAndColumns(ctx, first, client)
+	if err != nil {
+		return false, err
+	}
+	secondRows, secondColumns, err := rowsAndColumns(ctx, second, client)
+	if err != nil {
+		return false, err
+	}
+
+	logrus.Infof("Checking row names...")
+	if diff := cmp.Diff(firstRows, secondRows); diff != "" {
+		logrus.Infof("\tRows differ (-first, +second): %s", diff)
+		diffed = true
+	} else {
+		logrus.Infof("\tAll %d rows match!", len(firstRows))
+	}
+
+	logrus.Infof("Checking column data...")
+	if diff := cmp.Diff(firstColumns, secondColumns); diff != "" {
+		logrus.Infof("Columns differ (-first, +second): %s", diff)
+		diffed = true
+	} else {
+		logrus.Infof("\tAll %d columns match!", len(firstColumns))
+	}
+	return diffed, nil
+}
+
+func filenames(ctx context.Context, dir gcs.Path, client gcs.Client) []string {
+	stats := client.Objects(ctx, dir, "/", "")
+	var filenames []string
+	for {
+		stat, err := stats.Next()
+		if err != nil {
+			break
+		}
+		filename := filepath.Join(dir.String(), filepath.Base(stat.Name))
+		filenames = append(filenames, filename)
+	}
+	return filenames
+}
+
+func main() {
+	opt := gatherOptions()
+	opt.validate()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	storageClient, err := gcs.ClientWithCreds(ctx, opt.creds)
+	if err != nil {
+		logrus.Fatalf("Failed to create storage client: %v", err)
+	}
+	defer storageClient.Close()
+
+	client := gcs.NewClient(storageClient)
+
+	firstFiles := filenames(ctx, opt.first, client)
+	// secondFiles := filenames(ctx, opt.second, client)
+	var diffedMsgs, errorMsgs []string
+	for _, firstP := range firstFiles {
+		secondP := filepath.Join(opt.second.String(), filepath.Base(firstP))
+		firstPath, err := gcs.NewPath(firstP)
+		if err != nil {
+			errorMsgs = append(errorMsgs, fmt.Sprintf("gcs.NewPath(%q): %v", firstP, err))
+			continue
+		}
+		secondPath, err := gcs.NewPath(secondP)
+		if err != nil {
+			errorMsgs = append(errorMsgs, fmt.Sprintf("gcs.NewPath(%q): %v", secondP, err))
+			continue
+		}
+		diffed, err := compare(ctx, *firstPath, *secondPath, client)
+		if err != nil {
+			errorMsgs = append(errorMsgs, fmt.Sprintf("compare(%q, %q): %v", firstP, secondP, err))
+			continue
+		}
+		if diffed {
+			diffedMsgs = append(diffedMsgs, fmt.Sprintf("%q vs. %q", firstP, secondP))
+		}
+	}
+	if len(errorMsgs) > 0 {
+		logrus.Errorf("Got %d errors:", len(errorMsgs))
```

Comment:
Consider something more like `logrus.WithField("count", len(errorMsgs)).Error("Directories are not the same")`

Also consider changing this to `logrus.Fatal()` and moving at the end so it exits non-zero on failure.